### PR TITLE
Pause Menu

### PR DESCRIPTION
**Pull request by AustinKellar**
Contributors: Alfredo, Austin

Goals:
1) Implement a method of pausing the game during a match
2) Implement an easy way to return to character select, quit the game and restart the match

Important Files:
1) MatchManager.cs 
    - When a player presses start, stop time and open _that player's_ pause menu
    - Exposed functions for Restart Match, Quit Game and Return to Player Select
    - Exposed a new variable called IsPaused
2) UIPauseMenu.cs
    - Controls the pause menu UI elements when IsPaused is true.
    - Calls exposed functions from MatchManager.cs
3) UIPauseMenuInput.cs
    - Handles controller / keyboard input logic for the pause menu
4) UIMatchManager.cs
    - Repurposed this file to generally encapsulate match UI instead of only Game Over text
5) UIGameOver.cs
    - Moved Game Over UI functionality from UIMatchManager.cs to this file

How to Test:
- IMPORTANT test in standalone build and Unity editor
- Load the scene testing/pause menu
- Press Start (escape on keyboard)
    - Make sure game time was paused
    - Make sure you can navigate the menu with up / down arrow keys, analog stick, dpad (inverted on PS4 controller)
    - Make sure that only the player who pressed pause can control the pause menu
    - Make sure you can exit menu with B and Start
    - Make sure that all menu options behave the way you expect when you press A (Right shift on keyboard)
        - Note: You can only Close Game if you are in standalone build
    - Make sure you hear sounds when you open the menu, close the menu and select an option

Notes: 
- We built our own manual hovering system instead of using Unity's built in navigation.
- In order to test, we changed the build settings to start the pause menu scene

### CODE REVIEW

File: Assets/Scripts/Managers/MatchManager.cs [hunk 1]
```diff
@@ -39,18 +48,72 @@ private void Start()
...
+            AudioManager.Instance.PlayOneShot("UI Close Menu");
+            Time.timeScale = 1f;
+            UIMatchManager.Instance.PauseMenuCanvas.SetActive(false);
+
+            Timer timer = null;
+            timer = new Timer((obj) =>
+            {
+                IsPaused = false;
+                timer.Dispose();
+            }, null, 200, Timeout.Infinite);
```
Comment (IanGRap) [inline Assets/Scripts/Managers/MatchManager.cs]: just curious, why do we need a timer to close the pause menu?

File: Assets/Scripts/Match UI/UIPauseMenu.cs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+    private void Start()
+    {
+        _actions = new List<MenuAction>
+        {
+            new MenuAction(() => { MatchManager.Instance.TogglePauseGame(PlayerIndex.Any); }, _resumeButton),
+            new MenuAction(MatchManager.Instance.RestartMatch, _restartButton),
+            new MenuAction(MatchManager.Instance.LoadPlayerSelect, _playerSelectButton),
+            new MenuAction(MatchManager.Instance.QuitApplication, _quitButton)
+        };
```
Comment (IanGRap) [inline Assets/Scripts/Match UI/UIPauseMenu.cs]: sexy

File: Assets/Scripts/Match UI/UIPauseMenu.cs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        {
+            AudioManager.Instance.PlayOneShot("UI Menu Confirm");
+            _selectedAction.action.Invoke();
+        }
+
+        _actions.ForEach(a => 
+        {
+            a.button.GetComponent<Image>().color = _staticColor;
+            a.button.transform.localScale = Vector3.one;
+        });
```
Comment (IanGRap) [inline Assets/Scripts/Match UI/UIPauseMenu.cs]: The for loop seems a little unnecessary to me if you just tracked previous action but 4 actions isn't enough for me to worry about performance at all

File: Assets/Scripts/Managers/MatchManager.cs (hunk 1 above)
Comment (AustinKellar) [reply Assets/Scripts/Managers/MatchManager.cs]: Without the timer, if you press A to resume, your character will immediately jump. This is just to stop that from happening.

File: Assets/Scripts/Match UI/UIPauseMenu.cs (hunk 2 above)
Comment (AustinKellar) [reply Assets/Scripts/Match UI/UIPauseMenu.cs]: tyty

File: Assets/Scripts/Match UI/UIPauseMenu.cs (hunk 3 above)
Comment (AustinKellar) [reply Assets/Scripts/Match UI/UIPauseMenu.cs]: Yeah, if we hit performance problems, I'll be happy to change it

**Pull request merged**